kame/kame/pvcsif/pvcsif.8
.Dd December 29, 1997
.Dt PVCSIF 8
.Os BSD
.Sh NAME
.Nm pvcsif
.Nd configure ATM PVC sub interfaces
.Sh SYNOPSIS
.Nm pvcsif
.Ar interface
.Op Fl s
.Nm pvcsif
.Fl a
.Sh DESCRIPTION
.Nm
creates a sub interface for an ATM PVC.
A sub interface 
.Xr pvc 4
is dynamically created.
The created interface is bound to 
.Ar interface
but at this point no VC is assigned.  To assign a VC, 
.Xr pvctxctl 8
should be used later. 
.Pp
A PVC sub interface is intended to use an ATM PVC as an alternative
serial connection, and to be allocated per PVC basis.
A PVC sub interface looks as a point-to-point interface
and is multicast capable, as opposed to the NBMA 
(NonBroadcast Multiple Access) model that requires a MARS server.
A point-to-point interface is useful to run MBone or protocols
requiring multicast, such as RSVP and IPv6, over a PVC WAN
connection.
.Pp
Note that a sub interface is not a full-fledged interface but just
an indirect reference to the real interface.
.Pp
The options are as follows:
.Bl -tag -width Fl
.It Fl s
For use with a shell, it prints the created interface name.
.It Fl a
Lists the existing sub interfaces.
.El
.Sh BUGS
Currently,  there is no way to remove a sub interface.
.Sh SEE ALSO
.Xr pvctxctl 8 , 
.Xr ifconfig 8 , 
.Xr en 4


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/pvcsif/pvcsif.c
/*
 * Copyright (C) 1998
 *	Sony Computer Science Laboratory Inc.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY SONY CSL AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL SONY CSL OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id: pvcsif.c,v 1.2 2001/05/08 04:36:35 itojun Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <err.h>

#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/ioctl.h>
#include <net/if.h>

#include <net/if_atm.h>

#include "if_nameindex.h"

void list_all(void);

void usage(void)
{
	fprintf(stderr, "usage: pvcsif -a\n");
	fprintf(stderr, "       pvcsif interface [-s]\n");
	exit(1);
}

int 
main(int argc, char **argv)
{
	struct ifreq ifr;
	int shell_mode = 0;
	int s, ch;

	if (argc < 2)
		usage();

	if (strncmp(argv[1], "-a", 2) == 0) {
		list_all();
		return (0);
	}

	ifr.ifr_name[IFNAMSIZ-1] = '\0';
	strncpy(ifr.ifr_name, argv[1], IFNAMSIZ-1);

	optind = 2;
	while ((ch = getopt(argc, argv, "s")) != -1) {
		switch (ch) {
		case 's':
			shell_mode = 1;
			break;
		}
	}

	if ((s = socket(AF_INET, SOCK_DGRAM, 0)) < 0)
		err(1, "can't open socket");

	if (ioctl(s, SIOCSPVCSIF, &ifr) < 0)
		err(1, "SIOCSPVCSIF");

	close(s);

	if (shell_mode)
		printf("%s", ifr.ifr_name);
	else
		printf("created a pvc subinterface %s (bound to %s)\n",
		       ifr.ifr_name, argv[1]);
    
	return (0);
}

void
list_all(void)
{
	struct if_nameindex *ifn_list, *ifnp;
	struct pvctxreq pvcreq;
	int pcr, s;

	if ((ifn_list = if_nameindex()) == NULL)
		err(1, "if_nameindex failed");

	if ((s = socket(AF_INET, SOCK_DGRAM, 0)) < 0)
		err(1, "can't open socket");

	for (ifnp = ifn_list; ifnp->if_name != NULL; ifnp++) {
		if (strncmp(ifnp->if_name, "pvc", 3) == 0) {

			bzero(&pvcreq, sizeof(pvcreq));
			strncpy(pvcreq.pvc_ifname, ifnp->if_name, IFNAMSIZ-1);
			if (ioctl(s, SIOCGPVCTX, &pvcreq) < 0)
				err(1, "SIOCSPVCTX");

			/*
			 * print info
			 */
			pcr = pvcreq.pvc_pcr;
			printf("  %s (bound to %s): vci:[%d:%d] (",
			       ifnp->if_name, pvcreq.pvc_ifname,
			       ATM_PH_VPI(&pvcreq.pvc_aph),
			       ATM_PH_VCI(&pvcreq.pvc_aph));
			if (ATM_PH_FLAGS(&pvcreq.pvc_aph) & ATM_PH_AAL5)
				printf("AAL5");
			if (ATM_PH_FLAGS(&pvcreq.pvc_aph) & ATM_PH_LLCSNAP)
				printf("/LLCSNAP");
			printf(") ");
			if (pcr < 0)
				printf("(invalid)\n");
			else if (pcr == 0)
				printf("pcr:%d(full speed)\n", pcr);
			else if (pcr < 1000)
				printf("pcr:%d(%dbps)\n",
				       pcr, pcr * 48 * 8);
			else if (pcr < 1000000)
				printf("pcr:%d(%dKbps)\n",
				       pcr, pcr * 48 * 8 / 1000);
			else
				printf("pcr:%d(%dMbps)\n",
				       pcr, pcr * 48 * 8 / 1000000);
		}
	}

	close(s);
	if_freenameindex(ifn_list);
}
